tensorflow/compiler/xla/service/gpu/ir_emission_utils.cc

```diff
@@ -319,10 +319,17 @@ llvm::Value* EmitFullWarpShuffleDown(llvm::Value* value, llvm::Value* offset,
 
   // Special case for efficiency
   if (value->getType()->isFloatTy() && bit_width == 32) {
-    return EmitCallToTargetIntrinsic(
-        TargetIntrinsicID::kShflDownF32,
-        {all_warps_mask, value, offset, builder->getInt32(kWarpSize - 1)}, {},
-        builder);
+    struct TargetFunctionCallInfo target_function_call_info(
+        TargetFunctionID::kShflDownF32);
+    // Input operands.
+    target_function_call_info.operands = {all_warps_mask, value, offset,
+                                          builder->getInt32(kWarpSize - 1)};
+    // Operand types.
+    target_function_call_info.input_types = {S32, F32, S32, S32};
+    // Result type of the device function.
+    target_function_call_info.output_type = F32;
+    // Emit intrinsic or device function.
+    return EmitCallToTargetFunction(target_function_call_info, builder);
```

Comment:
Do we only need `struct TargetFunctionCallInfo` for shfldown?  In that case, could we just move the nvidia/amd-specific logic into this file, and remove a bunch of complexity / generality?  Right now this is pretty confusing because we still are exposing in an API the fact that shfl takes four arguments, except on AMD only two of them are used, etc etc.

As a concrete example, if you'd written `input_types = {S32, S32, F32, S32}` instead of what's above, I'd have had no idea that this was wrong.

Reply:
You mean it would be ok to check for if the target is  PTX or AMDGPU inside ir_emission_utils.cc ? 

> github's code review UI is having a serious problem with this PR; I'm unable to leave comments on most lines. (When I try to, I get an HTTP 422 response back from clicking the "add comment" button.) I've sent them a bug report.
> 
> In the meantime, I will try to put comments in here:
> 
> ```
>     llvm::FunctionType* callee_type = llvm::FunctionType::get(
>         /*Result=*/llvm_ir::PrimitiveTypeToIrType(output_type,
>                                        module),
> ```
> 
> This line and everything below it is unreachable code in this current patch, right?

You mean because there is no AMDGPU Target ? Yes, that is true though locally we do exercise it. I can remove it if thats desired at this point and reintroduce later. 


> If we go with the suggestion of moving the shfl code entirely into ir_emission_utils (thus eliminating this generality that doesn't seem to generalize?) then we probably don't need the `TargetFunctionCallInfo` struct either. This lets us keep `EmitCallToTargetFunction` for those functions which behave identically (or at least, close to the same) on both AMDGPU and NVPTX.

Wont that then introduce PTX/AMDGPU target  checking inside ir_emission_utils ?  I

Reply:
> Wont that then introduce PTX/AMDGPU target checking inside ir_emission_utils ?

Yes.  This is what I was trying to get at earlier when I said:

> To me, it seems that TargetUtil should be used for cases where there are NVPTX and AMDGCN functions that are basically equivalent. In this case they're not, because the NVPTX version takes additional args.
>
> Instead, could we have an "emit shfl down" function which emits the two-arg version on amdgpu and the four-arg version on nvptx? This function might use TargetUtil, or it might not, because it's already switch'ing on the target anyway.

>> This line and everything below it is unreachable code in this current patch, right?
>
> You mean because there is no AMDGPU Target ?

I mean because that code only applies for calls implemented by functions (i.e. not intrinsics), and it does not apply to AMDGPU shfl.  But, at least today, the only calls implemented by functions are AMDGPU shfl.

Reply:
Ok..I have removed the implementation for emitting device functions.